### PR TITLE
Feature/add remote location

### PR DESCRIPTION
**Pull request by urielfcampos**
Hey ! 

I've added the support to having members register as remote workers it work as following:

`/register_remote_member [country_from] [country_to]` And it will register a new user that's working remotely from one country to another, or even to the same country.

`/deregister_remote_member` removes a user from the remote workers list

`/ping_remote` pings all remote users

`/find_remote_member_from [country_code]` finds remote workers from a certain country

`/find_remote_country_to [country_code]`  finds remote workers working to a certain country

The last two commands are still to be done, and i hope to finish them tomorrow.

Let me know what you think !

### CODE REVIEW

File: src/command.ts [hunk 1]
```diff
@@ -50,6 +60,14 @@ const DefaultCommandDescriptions: Record<Command, string> = {
...
   [Command.FindMember]: 'Find all locations you are registered to.',
   [Command.Help]: 'Help about various subjects.',
+  [Command.FindRemoteMemberFrom]:
+    'Find members that are remote from a country',
+  [Command.FindRemoteMemberTo]:
+    'Find members that are remote to a country',
+  [Command.RegisterRemoteMember]: 'Register you as a remote member',
+  [Command.DeregisterRemoteMember]:
+    'Deregister you from the remote members',
+  [Command.PingRemote]: 'Ping remote remembers',
```
Comment (armand1m) [inline src/command.ts:70]:
```suggestion
  [Command.PingRemote]: 'Ping remote members',
```

File: src/utils/country.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import { BotContext } from '../context';
+import { getCountryCodeForText } from '../countries';
+
+export const validateCountry = (
+  countryName: string,
+  ctx: BotContext
+) => {
```
Comment (armand1m) [inline src/utils/country.ts:7]: this is quite a nice utility function, will be very handy to remove some duplication I see in the remote commands

**Review by armand1m — COMMENTED** [review]
some comments here and there but the overall work is pretty straightforward

thank you a lot for the contribution!

File: src/database.ts [hunk 3]
```diff
@@ -91,6 +112,55 @@ export const createDatabase = async (
...
+    getRemoteMembersTo: (countryCode) => {
+      const collection = db.get('remoteIndex')
+      const members = collection.value() || [];
+
+      const filteredMembers = Object.entries(members).filter(([key, value]) => {
+        if (value["to"] !== undefined) {
+          return value["to"] === countryCode
+        }
+        return false
+      })
```
Comment (armand1m) [inline src/database.ts:155]: this expression could be simplified to just be `return value["to"] === countryCode`

File: src/database.ts [hunk 4]
```diff
@@ -91,6 +112,55 @@ export const createDatabase = async (
...
+    getRemoteMembersFrom: (countryCode) => {
+      const collection = db.get('remoteIndex')
+      const members = collection.value() || [];
+
+      const filteredMembers = Object.entries(members).filter(([key, value]) => {
+        if (value["from"] !== undefined) {
+          return value["from"] === countryCode
+        }
+        return false
+      })
```
Comment (armand1m) [inline src/database.ts:142]: this expression could be simplified to just be `return value["from"] === countryCode`

File: src/middlewares/createContextMiddleware.ts [hunk 5]
```diff
@@ -128,7 +128,43 @@ export const createContextMiddleware = ({ config }: Props) => {
...
+              err.message.includes('user not found')
+            ) {
+              ctx.logger.warn(
+                `Registered user with id "${userId}" does not exist. Removing user remote.`
+              );
+              ctx.database.removeRemoteMember(Number(userId));
+            }
+            throw err;
+          }
+        })
```
Comment (armand1m) [inline src/middlewares/createContextMiddleware.ts:153]: probably would be nice to have this function extracted into it's own:

```ts
const fetchMemberMention = async (userId: number, silenced: boolean) => {
  try {
    const member = await ctx.getChatMember(Number(userId));
    return createMemberMention(member.user, silenced);
  } catch (err: any) {
    if (
      err.code === 400 &&
      err.message.includes('user not found')
    ) {
      ctx.logger.warn(
        `Registered user with id "${userId}" does not exist. Removing user remote.`
      );
      ctx.database.removeRemoteMember(Number(userId));
    }
    throw err;
  }
}
```

and then just use it in the `fetchMembersMentionList` and `fetchRemoteMembersMentionList`:

```ts
// fetchMembersMentionList version
const membersFetchResult = await Promise.allSettled(
  memberIds.map((userId) => {
    return fetchMemberMention(userId, silenced)
  })
);

// fetchRemoteMembersMentionList version
const membersFetchResult = await Promise.allSettled(
  Object.keys(allMembers).map((userId) => {
    return fetchMemberMention(userId, silenced)
  })
);
```

**Review by armand1m — COMMENTED** [review]
added some code recommendations

File: src/middlewares/createContextMiddleware.ts (hunk 5 above)
Comment (urielfcampos) [reply src/middlewares/createContextMiddleware.ts:153]: Nice idea, will implement it !

File: src/database.ts (hunk 3 above)
Comment (urielfcampos) [reply src/database.ts:155]: That makes sense, it will simplify the code haha

File: src/utils/country.ts (hunk 2 above)
Comment (urielfcampos) [reply src/utils/country.ts:7]: Glad to hear it !

File: src/middlewares/createContextMiddleware.ts (hunk 5 above)
Comment (urielfcampos) [reply src/middlewares/createContextMiddleware.ts:124]: Done.

File: src/database.ts (hunk 4 above)
Comment (urielfcampos) [reply src/database.ts]: Done.

File: src/utils/country.ts (hunk 2 above)
Comment (urielfcampos) [reply src/utils/country.ts]: Refactored `registerRemoteMember.ts` to use this function.

File: src/command.ts (hunk 1 above)
Comment (urielfcampos) [reply src/command.ts]: Done haha

**Review by armand1m — APPROVED** [review]

**Pull request merged**